src/code/m_lights.c

```diff
@@ -0,0 +1,366 @@
+            } else if (kq < 20) {
+                kq = 20;
+            }
+
+            light->l.col[0] = lightInfo->point.color[0];
+            light->l.colc[0] = light->l.col[0];
+
+            light->l.col[1] = lightInfo->point.color[1];
+            light->l.colc[1] = light->l.col[1];
+
+            light->l.col[2] = lightInfo->point.color[2];
+            light->l.colc[2] = light->l.col[2];
+
+            light->p.pos[0] = lightInfo->point.x;
+            light->p.pos[1] = lightInfo->point.y;
+            light->p.pos[2] = lightInfo->point.z;
+
+            light->p.kc = 8;
+            light->p.kl = -1;
+            light->p.kq = (s32)kq;
+        }
+    }
+}
+#else
+void LightsN__P_point_proc(LightsN* lights, LightParams* lightInfo, UNUSED Vec3f* pos);
+GLOBAL_ASM("asm/jp/nonmatchings/code/m_lights/LightsN__P_point_proc.s")
+#endif
+
+void LightsN__diffuse_proc(LightsN* lights, LightParams* lightInfo, UNUSED Vec3f* pos) {
+    Light* light = LightsN_new_diffuse(lights);
+
+    if (light != NULL) {
+        light->l.col[0] = light->l.colc[0] = lightInfo->diffuse.color[0];
+        light->l.col[1] = light->l.colc[1] = lightInfo->diffuse.color[1];
+        light->l.col[2] = light->l.colc[2] = lightInfo->diffuse.color[2];
+        light->l.dir[0] = lightInfo->diffuse.x;
+        light->l.dir[1] = lightInfo->diffuse.y;
+        light->l.dir[2] = lightInfo->diffuse.z;
+    }
+}
+
+void LightsN_list_check(LightsN* lights, LightNode* node, Vec3f* pos) {
+    static light_point_proc poslight_type_proc[] = { LightsN__point_proc, LightsN__diffuse_proc, LightsN__point_proc };
+    static light_P_point_proc light_type_proc[] = { LightsN__P_point_proc, LightsN__diffuse_proc,
+                                                    LightsN__P_point_proc };
+
+    if (pos == NULL) {
+        while (node != NULL) {
+            light_type_proc[node->info->type](lights, &node->info->lights, pos);
+            node = node->next;
+        }
+    } else {
+        while (node != NULL) {
+            poslight_type_proc[node->info->type](lights, &node->info->lights, pos);
+            node = node->next;
+        }
+    }
+}
+
+LightNode* Light_list_buf_new(void) {
+    LightNode* lightNode;
+
+    if (light_list_buf.current >= ARRAY_COUNT(light_list_buf.lights)) {
+        return NULL;
+    }
+
+    lightNode = &light_list_buf.lights[light_list_buf.idx];
+
+    while (lightNode->info != NULL) {
+        light_list_buf.idx++;
+        if (light_list_buf.idx < ARRAY_COUNT(light_list_buf.lights)) {
+            lightNode++;
+        } else {
+            light_list_buf.idx = 0;
+            lightNode = &light_list_buf.lights[0];
+        }
+    }
+    light_list_buf.current++;
+    return lightNode;
+}
+
+void Light_list_buf_delete(LightNode* lightNode) {
+    if (lightNode != NULL) {
+        light_list_buf.current--;
+        lightNode->info = NULL;
+        light_list_buf.idx = (lightNode - light_list_buf.lights) / (s32)sizeof(LightNode);
```

Comment:
Is the `s32` cast required here?

Reply:
Yes, sizeof returns a u32 and it does divu rather than div without it.